weedcoco/repo/deposit.py

```diff
@@ -2,178 +2,325 @@
-    with open(weedcoco_path) as f:
-        weedcoco = json.load(f)
-    for image_reference in weedcoco["images"]:
-        file_name = image_reference["file_name"].split("/")[-1]
-        if file_name in image_hash:
-            image_reference["file_name"] = "images/" + image_hash[file_name]
-
-    validate(weedcoco, image_dir)
-    new_dataset_dir = dataset_dir / "weedcoco.json"
-    with (new_dataset_dir).open("w") as out:
-        json.dump(weedcoco, out, indent=4)
-
-
-def retrieve_image_paths(images_path):
-    for root, directories, files in os.walk(images_path):
-        for filename in files:
-            yield (os.path.join(root, filename), filename)
-
-
-def compress_to_download(dataset_dir, deposit_id, download_dir):
-    mkdir_safely(download_dir)
-    download_path = (download_dir / deposit_id).with_suffix(".zip")
-    with tempfile.TemporaryDirectory() as tmpdir:
-        zip_path = pathlib.Path(tmpdir) / "bundle.zip"
-        with ZipFile(zip_path, "w") as zip:
-            zip.write(dataset_dir / "weedcoco.json", "weedcoco.json")
-            for image, image_name in retrieve_image_paths(dataset_dir / "images"):
-                zip.write(image, "images/" + image_name)
-        # XXX: this should be move() not copy(), but move resulted in files
-        #      that we could not delete or move in the Docker volume.
-        copy(zip_path, download_path)
-
-
-def atomic_copy(repository_dir, download_dir, temp_dir, deposit_id):
-    dataset_path = temp_dir / deposit_id
-    zip_path = temp_dir / f"{deposit_id}.zip"
-    repository_dir = repository_dir / deposit_id
-    try:
-        if os.path.isfile(zip_path):
-            move(str(zip_path), str(download_dir))
-        if os.path.isdir(dataset_path):
-            move(str(dataset_path), str(repository_dir))
-    except Exception:
-        # roll back to leave a clean repository
-        (download_dir / zip_path).unlink(missing_ok=True)
-        for path in (repository_dir / deposit_id).glob("*"):
-            path.unlink(missing_ok=True)
-        (repository_dir / deposit_id).rmdir()
-        raise
-
-
-def deposit(weedcoco_path, image_dir, repository_dir, download_dir, upload_id=None):
-    image_hash = create_image_hash(image_dir)
-    validate_duplicate_images(image_hash)
-    validate_existing_images(repository_dir, image_hash)
-    with tempfile.TemporaryDirectory() as temp_dir:
-        temp_dir = pathlib.Path(temp_dir)
-        dataset_dir, deposit_id = setup_dataset_dir(repository_dir, temp_dir, upload_id)
-        deposit_weedcoco(weedcoco_path, dataset_dir, image_dir, image_hash)
-        migrate_images(dataset_dir, image_dir, image_hash)
-        compress_to_download(dataset_dir, deposit_id, temp_dir)
-        atomic_copy(repository_dir, download_dir, temp_dir, deposit_id)
-    return str(repository_dir)
+
+def deposit(
+    weedcoco_path, image_dir, repository_dir, download_dir, metadata, upload_id=None
```

Comment:
A minor thought: Several of these args could be determined from a single configuration object. These could be encapsulated in a config object (or something with the attributes of django's `core.settings`), or should at least be grouped together as args. I assume `weedcoco_path` incorporates the `upload_id` in practice??